contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,248 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
+* **RunAsGroup**: same as **RunAsUser** but for gid mapping ranges.
+* **RunAsNonRoot**: Today, kubelet fails to start the container if image is found to be having UID 0. After usernamespace support in kubelet, kubelet should ignore this check if runtime is found to be running with usernamespace remapping enabled. Kubelet should verify image UID only when it is learned that usernamespace remapping is either not supported or not enabled at runtime.
+* **FsGroup and SupplementalGroups**: FsGroup and SupplementalGroups will continue to work after usernamespace support is Kubelet if runtime either does not support usernamespaces or support is not enabled. In case usernamespace support is discovered by kubelet as enabled at runtime, FsGroups/SupplementalGroups will work only if supported by the runtime (CRI implementation). One implementation, for example, could be where runtime(CRI implementation) will add a triplet in `/proc/<container-pid>/gid_map` for the FsGroup/Supplemental GIDs. This will be a 1:1 mapping between host usernamespace group id and container usernamespace group id i.e container ID and host ID will be same. Example:
 
-- **Beta:** Explicit API to request host user-namespace in pod spec
-          This is being targeted under Beta graduation plans.
+    If pod has `FsGroup: 2000` and runtime's default gid mappings are `containerid 0,  hostid 100000, size 1000`, runtime should be adding two entries in the `/proc/<container-pid>/gid_map`:
+    - [0 100000 1000]  // From GID 0 to 1000 in the container are mapped with GID range 100k to 100k+1000 on host
+    - [2000 2000 1]    // GID 2000 in container usernamespace is mapped to GID 2000 is host usernamespace
 
-### CRI API Changes
-Proposed CRI API changes:
+    Once this 1:1 mapping is set for the container process, `runc` already has the logic to call `setgroups()` to set the FsGroupID as the supplemental ID in the container usernamespace.
 
-```golang
-// Runtime service defines the public APIs for remote container runtimes
-service RuntimeService {
+    Setting the permissions on volume is resposibility of volume plugin. Once GID on the volume matches the supplemental GID in the container, access will work.
+
+    NOTE: Pod will be failed admission at kubelet if FsGroup/SupplementalGroups are in the default mapping range and mappings are not 1:1 i.e host id for a container id is different.
+
+
+### Volume Permissions  
+Kubelet will change the file permissions, i.e chown, at `/var/lib/kubelet/pods/<pod-uid>/volumes` after volume plugin is done with provisioning to get updated ownership of the volume directory according to remapped UID and GID.
+This proposal will work only for local volumes and not with remote volumes which have client-server kind of complex architecture where client is not allowed to chown volumes as root such as NFS.
```

Comment:
Hmm, that seems sort of confusing. So, if I'm using user NS remapping, some volumes will be chown'ed, and others will be left as-is?

Reply:
If any of the volumes is found to be NFS volume, pod will be defaulted to use host usernamespace, like in case of host path volumes.

Reply:
I still think that's confusing. The user needs to know whether a given volume type respects user namespace remapping before they use it. That seems a especially problematic when CSI is thrown into the mix.